src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,489 @@
+            allocator,
+            options.blocks_max + options.tables_max * lsm_table_content_blocks_max,
+        );
+
+        return GridRepairQueue{
+            .options = options,
+            .faulty_blocks = faulty_blocks,
+        };
+    }
+
+    pub fn deinit(queue: *GridRepairQueue, allocator: std.mem.Allocator) void {
+        assert(!queue.canceling);
+
+        queue.faulty_blocks.deinit(allocator);
+
+        queue.* = undefined;
+    }
+
+    /// When the queue wants more blocks than fit in a single request message, successive calls
+    /// to this function cycle through the pending BlockRequests.
+    pub fn block_requests(queue: *GridRepairQueue, requests: []vsr.BlockRequest) usize {
+        assert(!queue.canceling);
+
+        const faults_total = queue.faulty_blocks.count();
+        if (faults_total == 0) return 0;
+        assert(faults_total > queue.faulty_blocks_repair_index);
+
+        const fault_addresses = queue.faulty_blocks.entries.items(.key);
+        const fault_data = queue.faulty_blocks.entries.items(.value);
+
+        var requests_count: usize = 0;
+        var fault_offset: usize = 0;
+        while (requests_count < requests.len and
+            fault_offset < faults_total) : (fault_offset += 1)
+        {
+            const fault_index =
+                (queue.faulty_blocks_repair_index + fault_offset) % faults_total;
+
+            const fault = &fault_data[fault_index];
+            if (fault.state == .waiting) {
+                requests[requests_count] = .{
+                    .block_address = fault_addresses[fault_index],
+                    .block_checksum = fault_data[fault_index].checksum,
+                };
+                requests_count += 1;
+            }
+        }
+
+        queue.faulty_blocks_repair_index =
+            (queue.faulty_blocks_repair_index + fault_offset) % faults_total;
+        return requests_count;
+    }
+
+    /// Count the number of *non-table* block repairs available.
+    pub fn request_blocks_available(queue: *const GridRepairQueue) usize {
+        assert(!queue.canceling);
+        assert(queue.faulty_tables.count <= queue.options.tables_max);
+
+        const faulty_blocks_free =
+            queue.faulty_blocks.capacity() -
+            queue.faulty_blocks.count() -
+            queue.options.tables_max * lsm_table_content_blocks_max;
+        return faulty_blocks_free;
+    }
+
+    /// Queue a faulty block to request from the cluster and repair.
+    pub fn request_block(queue: *GridRepairQueue, address: u64, checksum: u128) void {
+        assert(!queue.canceling);
+        assert(queue.request_blocks_available() > 0);
+        assert(queue.faulty_tables.count <= queue.options.tables_max);
+
+        if (queue.faulty_blocks.getPtr(address)) |fault| {
+            assert(fault.checksum == checksum);
+            assert(fault.state != .aborting);
+            return;
+        }
+
+        queue.faulty_blocks.putAssumeCapacityNoClobber(address, .{
```

Comment:
could `get_or_put` here to avoid two hash map lookups. I think we also have three call-sites where we add entries for faulty_blocks (request{_table,_table_content,_block}). Maybe it makes sense to extract a helper such that there's only a single place where we physically add block to the hash map, so that there's a single place where we check validity of all transitions (eg, if we concurrently request the block for different reasons).